spyder/widgets/editor.py

```diff
@@ -0,0 +1,549 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright © Spyder Project Contributors
+#
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+# ----------------------------------------------------------------------------
+
+"""
+Simple code editor with syntax highlighting and line number area.
+
+Adapted from:
+https://doc.qt.io/qt-5/qtwidgets-widgets-codeeditor-example.html
+"""
+
+# Standard library imports
+import re
+
+# Third party imports
+from qtpy.QtCore import QRect, QSize, Qt, Signal
+from qtpy.QtGui import QColor, QPainter, QTextCursor, QTextFormat, QTextOption
+from qtpy.QtWidgets import QApplication, QPlainTextEdit, QTextEdit, QWidget
+
+# Local imports
+import spyder.utils.syntaxhighlighters as sh
+from spyder.widgets.mixins import BaseEditMixin
+
+
+# Constants
+LANGUAGE_EXTENSIONS = {
+    'Python': ('py', 'pyw', 'python', 'ipy'),
+    'Cython': ('pyx', 'pxi', 'pxd'),
+    'Enaml': ('enaml',),
+    'Fortran77': ('f', 'for', 'f77'),
+    'Fortran': ('f90', 'f95', 'f2k', 'f03', 'f08'),
+    'Idl': ('pro',),
+    'Diff': ('diff', 'patch', 'rej'),
+    'GetText': ('po', 'pot'),
+    'Nsis': ('nsi', 'nsh'),
+    'Html': ('htm', 'html'),
+    'Cpp': ('c', 'cc', 'cpp', 'cxx', 'h', 'hh', 'hpp', 'hxx'),
+    'OpenCL': ('cl',),
+    'Yaml': ('yaml', 'yml'),
+    'Markdown': ('md', 'mdw'),
+    # Every other language
+    'None': ('', ),
+}
+
+
+class LineNumberArea(QWidget):
+    """
+    Adapted from:
+    https://doc.qt.io/qt-5/qtwidgets-widgets-codeeditor-example.html
+    """
+
+    def __init__(self, code_editor=None):
+        super().__init__(code_editor)
+
+        self._editor = code_editor
+        self._left_padding = 6  # Pixels
+        self._right_padding = 3  # Pixels

```

Comment:
`_left_padding` and `_right_padding` are used directly below to set them, so I don't think there's a need to make them start with and underscore `_`.

Also, is there a need to make `_editor` a private attribute?

Reply:
>so I don't think there's a need to make them start with and underscore

If it is not meant to be exposed to the user it should have an underscore.

>Also, is there a need to make _editor a private attribute?

Cleanliness. The line number area accesses the editor internally so there is no reason why this should be exposed.

Reply:
Ok